File: src/domain/defaults.ts
/**
 * Default catalog provisioned for every new project: the named agents from
 * §8, the built-in MCP connections, default environments, the plan-mode
 * skill, and the `compound-engineer-workflow` template from §10.
 *
 * Reconstructed prompts come from src/prompts/prompts.ts (labeled there).
 */
import { randomUUID } from "node:crypto";
import {
  FOUNDATIONAL_PROMPT,
  ROLE_PROMPTS,
  PLAN_MODE_SKILL_BODY,
} from "../prompts/prompts.js";
import { DEFAULT_MODELS } from "../runners/models.js";
import type {
  Agent,
  Environment,
  Skill,
  McpConnection,
  TaskTemplate,
  TemplateStep,
  RunnerPreference,
} from "./types.js";

export interface ProjectDefaults {
  agents: Omit<Agent, "id" | "createdAt">[];
  environments: Omit<Environment, "id">[];
  skills: Omit<Skill, "id">[];
  mcpConnections: Omit<McpConnection, "id">[];
  templates: Omit<TaskTemplate, "id">[];
}

const plannerModel = DEFAULT_MODELS.planner;
const workerModel = DEFAULT_MODELS.worker;

/** Shared per-agent filesystem grants: own home folder + shared goal folder. */
const homeGrant = (slug: string) => ({
  folderPath: `/agents/${slug}`,
  canRead: true,
  canWrite: true,
  canDelete: true, // own home folder: delete allowed
});
const goalGrant = {
  folderPath: `/goals/`,
  canRead: true,
  canWrite: true,
  canDelete: false, // shared folder: write yes, delete still explicit (blueprint §7)
};
const wikiGrant = {
  folderPath: `/wiki`,
  canRead: true,
  canWrite: true,
  canDelete: false,
};

export function projectDefaults(projectId: string): ProjectDefaults {
  // These are environment NAMES — agent rows store them and ProjectService
  // resolves them against the seeded environments below (§5.5 network wall).
  const envNone = "limited-none";
  const envOpen = "open";

  const mcpAgentos = "agentos";
  const mcpInbox = "inbox";
  const mcpR2 = "r2-fs";
  const mcpGithub = "github";
  const mcpFront = "front";

  const environments: Omit<Environment, "id">[] = [
    {
      projectId,
      // Customer-support is "limited to Front" (§5.3): the wall must let the
      // support agent reach its one connection host and nothing else.
      name: "limited-none",
      networking: "limited",
      allowedHosts: ["api.front.com"],
      envNames: [],
    },
    {
      projectId,
      name: "open",
      networking: "open",
      allowedHosts: [],
      envNames: [],
    },
  ];

  const skills: Omit<Skill, "id">[] = [
    {
      projectId,
      name: "Plan mode",
      slug: "plan-mode",
      kind: "prompt",
      body: PLAN_MODE_SKILL_BODY,
      filePath: null,
    },
  ];

  const mcpConnections: Omit<McpConnection, "id">[] = [
    { projectId, name: "agentos", config: { builtin: true }, credentialSecretId: null },
    { projectId, name: "inbox", config: { builtin: true }, credentialSecretId: null },
    { projectId, name: "r2-fs", config: { builtin: true }, credentialSecretId: null },
    // External connections are configurable, not hardcoded product logic (§23).
    { projectId, name: "github", config: { provider: "github", url: "https://api.github.com" }, credentialSecretId: null },
    { projectId, name: "front", config: { provider: "front", url: "https://api.front.com" }, credentialSecretId: null },
  ];

  const agent = (
    name: string,
    title: string,
    model: string,
    opts: {
      role?: string;
      skills?: string[];
      mcp?: string[];
      repos?: Agent["repoAccess"];
      grants?: Agent["filesystemGrants"];
      collab?: string[];
      env?: string;
      runner?: RunnerPreference;
      inbox?: boolean;
    } = {},
  ): Omit<Agent, "id" | "createdAt"> => ({
    projectId,
    name,
    title,
    model,
    foundationalPrompt: FOUNDATIONAL_PROMPT,
    rolePrompt: opts.role ?? ROLE_PROMPTS[name] ?? ROLE_PROMPTS.default,
    skillIds: opts.skills ?? [],
    mcpConnectionIds: opts.mcp ?? [mcpAgentos, mcpInbox, mcpR2],
    repoAccess: opts.repos ?? [],
    filesystemGrants: opts.grants ?? [homeGrant(name), goalGrant],
    collaborationList: opts.collab ?? [],
    environmentId: opts.env ?? envNone,
    runnerPreference: opts.runner ?? "inherit",
    inboxAccess: opts.inbox ?? false,
  });

  const agents: Omit<Agent, "id" | "createdAt">[] = [
    agent("default", "Default workhorse", workerModel, { env: envOpen }),
    agent("plan", "Plan agent", plannerModel, {
      role: ROLE_PROMPTS.plan,
      skills: ["skill-plan-mode"],
      mcp: [mcpAgentos, mcpInbox, mcpR2],
      runner: "cloud",
    }),
    agent("spec", "Spec agent", plannerModel, {
      role: ROLE_PROMPTS.spec,
      runner: "cloud",
      inbox: true,
    }),
    agent("senior-dev", "Senior developer", workerModel, {
      role: ROLE_PROMPTS["senior-dev"],
      mcp: [mcpAgentos, mcpInbox, mcpR2, mcpGithub],
      env: envOpen,
    }),
    agent("implementation-plan-executioner", "Implementation plan executioner", workerModel, {
      role: ROLE_PROMPTS["implementation-plan-executioner"],
      mcp: [mcpAgentos, mcpInbox, mcpR2, mcpGithub],
      env: envOpen,
    }),
    agent("review-coordinator", "Review coordinator", plannerModel, {
      role: ROLE_PROMPTS["review-coordinator"],
      collab: ["feasibility", "scope-guardian", "coherence", "plan-risk"],
      runner: "cloud",
    }),
    agent("feasibility", "Feasibility reviewer", plannerModel, {
      role: ROLE_PROMPTS.feasibility,
      runner: "cloud",
    }),
    agent("scope-guardian", "Scope guardian", plannerModel, {
      role: ROLE_PROMPTS["scope-guardian"],
      runner: "cloud",
    }),
    agent("coherence", "Coherence reviewer", plannerModel, {
      role: ROLE_PROMPTS.coherence,
      runner: "cloud",
    }),
    agent("plan-risk", "Plan risk reviewer", plannerModel, {
      role: ROLE_PROMPTS["plan-risk"],
      runner: "cloud",
    }),
    agent("librarian", "Librarian", workerModel, {
      role: ROLE_PROMPTS.librarian,
      grants: [homeGrant("librarian"), wikiGrant],
      env: envOpen,
    }),
    // Optional-but-seeded extras named in §8.
    agent("customer-support", "Customer support", workerModel, {
      role: ROLE_PROMPTS["customer-support"],
      mcp: [mcpAgentos, mcpInbox, mcpFront], // Front only — no GitHub, no Gmail (§5.3)
      env: envNone,
      inbox: true,
    }),
    agent("diagnostic", "Bug diagnostic", workerModel, {
      role: ROLE_PROMPTS.diagnostic,
      mcp: [mcpAgentos, mcpInbox, mcpR2, mcpGithub],
      env: envOpen,
    }),
    agent("linkedin-content", "LinkedIn content", workerModel, {
      role: ROLE_PROMPTS["linkedin-content"],
      env: envNone,
    }),
  ];

  // Template steps: each step carries a tight brief — the deliverable and
  // the exit criteria — so the assigned agent knows what "done" means.

  const tpl: Omit<TaskTemplate, "id"> = {
    projectId,
    name: "compound-engineer-workflow",
    description:
      "~3-hour fully managed feature build (his words; actual runs ~5–6h): spec (human-gated) → plan → multi-agent plan review → revise → implement with E2E → code review → fixes → wiki → human PR review.",
    variables: ["branchName", "featureTitle"],
    steps: compoundEngineerSteps(),
  };

  return { agents, environments, skills, mcpConnections, templates: [tpl] };
}

export function newProjectId(): string {
  return randomUUID();
}

/**
 * §10 — compound-engineer-workflow steps (canonical prompts).
 * Shared by the default catalog and the prompt migration so both use the
 * exact same briefs. Each step keeps its {{variables}} for interpolation.
 */
export function compoundEngineerSteps(): TemplateStep[] {
  return [
    {
      name: "Write a spec",
      agentName: "spec",
      prompt:
        "Write a detailed feature specification for {{featureTitle}} (branch {{branchName}}). " +
        "Deliverable: a spec file attached to the task with context, requirements, acceptance criteria and an out-of-scope list. " +
        "Exit: the spec is attached and awaiting human approval — do not mark the task done.",
      approvalGate: true,
    },
    {
      name: "Plan",
      agentName: "plan",
      prompt:
        "Turn the approved spec for {{featureTitle}} (branch {{branchName}}) into a concrete, ordered implementation plan. " +
        "Deliverable: a plan file attached to the task. " +
        "Exit: the plan is attached and the task is finished. Do not implement.",
      approvalGate: false,
    },
    {
      name: "Plan review",
      agentName: "review-coordinator",
      prompt:
        "Spawn the plan-review specialists (feasibility, scope-guardian, coherence, plan-risk) for the plan of {{featureTitle}}. " +
        "Deliverable: a consolidated report with must-fix / should-fix items, attached to the task. " +
        "Exit: every specialist reported and the consolidation is attached. Do not fix anything yourself.",
      approvalGate: false,
    },
    {
      name: "Revise plan",
      agentName: "plan",
      prompt:
        "Revise the plan for {{featureTitle}} using the consolidated plan review (must-fix, then should-fix). " +
        "Deliverable: the revised plan attached to the task. " +
        "Exit: every must-fix is addressed in the revised plan and the task is finished.",
      approvalGate: false,
    },
    {
      name: "Implementation",
      agentName: "implementation-plan-executioner",
      prompt:
        "Implement {{featureTitle}} on branch {{branchName}} exactly per the revised plan. " +
        "Deliverable: a committed implementation. " +
        "Exit: all plan steps implemented, the repo's E2E tests run and surfaced issues fixed, work committed, task finished.",
      approvalGate: false,
    },
    {
      name: "Code review",
      agentName: "review-coordinator",
      prompt:
        "Spawn the code-review specialists for the implementation of {{featureTitle}} on branch {{branchName}}. " +
        "Deliverable: a consolidated report with must-fix / should-fix items, attached to the task. " +
        "Exit: every specialist reported and the consolidation is attached. Do not fix anything yourself.",
      approvalGate: false,
    },
    {
      name: "Apply review fixes",
      agentName: "senior-dev",
      prompt:
        "Apply the consolidated code-review fixes (must-fix, then should-fix) for {{featureTitle}} on branch {{branchName}}. " +
        "Deliverable: committed fixes. " +
        "Exit: must-fix items resolved, E2E re-run and passing, work committed, task finished.",
      approvalGate: false,
    },
    {
      name: "Librarian",
      agentName: "librarian",
      prompt:
        "Update the internal wiki to reflect how the codebase actually works after {{featureTitle}} on branch {{branchName}}. " +
        "Deliverable: wiki pages describing the post-change state. " +
        "Exit: wiki updated, task finished. Do not change product code.",
      approvalGate: false,
    },
    {
      name: "Human PR review",
      agentName: "human",
      prompt:
        "Review the PR for {{featureTitle}} on branch {{branchName}}; merge when satisfied.",
      approvalGate: true,
    },
  ];
}

File: agentos/domain/defaults.py
"""
Default project catalog. Port of src/domain/defaults.ts.
Agents, environments, skills, MCP connections, and the compound-engineer-workflow template.
"""
from agentos.prompts.prompts import FOUNDATIONAL_PROMPT, PLAN_MODE_SKILL_BODY, ROLE_PROMPTS
from agentos.runners.models import DEFAULT_MODELS


def _home_grant(slug: str) -> dict:
    return {"folderPath": f"/agents/{slug}", "canRead": True, "canWrite": True, "canDelete": True}


_GOAL_GRANT = {"folderPath": "/goals/", "canRead": True, "canWrite": True, "canDelete": False}
_WIKI_GRANT = {"folderPath": "/wiki", "canRead": True, "canWrite": True, "canDelete": False}

_PLANNER = DEFAULT_MODELS["planner"]
_WORKER = DEFAULT_MODELS["worker"]


def project_defaults(project_id: str) -> dict:
    env_none = "limited-none"
    env_open = "open"

    environments = [
        {"project_id": project_id, "name": "limited-none", "networking": "limited",
         "allowed_hosts": ["api.front.com"], "env_names": []},
        {"project_id": project_id, "name": "open", "networking": "open",
         "allowed_hosts": [], "env_names": []},
    ]

    skills = [
        {"project_id": project_id, "name": "Plan mode", "slug": "plan-mode",
         "kind": "prompt", "body": PLAN_MODE_SKILL_BODY, "file_path": None},
    ]

    mcp_connections = [
        {"project_id": project_id, "name": "agentos", "config": {"builtin": True}, "credential_secret_id": None},
        {"project_id": project_id, "name": "inbox", "config": {"builtin": True}, "credential_secret_id": None},
        {"project_id": project_id, "name": "r2-fs", "config": {"builtin": True}, "credential_secret_id": None},
        {"project_id": project_id, "name": "github", "config": {"provider": "github", "url": "https://api.github.com"}, "credential_secret_id": None},
        {"project_id": project_id, "name": "front", "config": {"provider": "front", "url": "https://api.front.com"}, "credential_secret_id": None},
    ]

    def agent(name, title, model, role=None, mcp=None, grants=None, collab=None,
              env=None, runner="inherit", inbox=False, skills_list=None):
        return {
            "project_id": project_id,
            "name": name, "title": title, "model": model,
            "foundational_prompt": FOUNDATIONAL_PROMPT,
            "role_prompt": role or ROLE_PROMPTS.get(name, ROLE_PROMPTS["default"]),
            "skill_ids": skills_list or [],
            "mcp_connection_ids": mcp or ["agentos", "inbox", "r2-fs"],
            "repo_access": [],
            "filesystem_grants": grants or [_home_grant(name), _GOAL_GRANT],
            "collaboration_list": collab or [],
            "environment_id": env or env_none,
            "runner_preference": runner,
            "inbox_access": inbox,
        }

    agents = [
        agent("default", "Default workhorse", _WORKER, env=env_open),
        agent("plan", "Plan agent", _PLANNER, role=ROLE_PROMPTS["plan"],
              skills_list=["skill-plan-mode"], runner="cloud"),
        agent("spec", "Spec agent", _PLANNER, role=ROLE_PROMPTS["spec"], runner="cloud", inbox=True),
        agent("senior-dev", "Senior developer", _WORKER, role=ROLE_PROMPTS["senior-dev"],
              mcp=["agentos", "inbox", "r2-fs", "github"], env=env_open),
        agent("implementation-plan-executioner", "Implementation plan executioner", _WORKER,
              role=ROLE_PROMPTS["implementation-plan-executioner"],
              mcp=["agentos", "inbox", "r2-fs", "github"], env=env_open),
        agent("review-coordinator", "Review coordinator", _PLANNER,
              role=ROLE_PROMPTS["review-coordinator"],
              collab=["feasibility", "scope-guardian", "coherence", "plan-risk"], runner="cloud"),
        agent("feasibility", "Feasibility reviewer", _PLANNER, role=ROLE_PROMPTS["feasibility"], runner="cloud"),
        agent("scope-guardian", "Scope guardian", _PLANNER, role=ROLE_PROMPTS["scope-guardian"], runner="cloud"),
        agent("coherence", "Coherence reviewer", _PLANNER, role=ROLE_PROMPTS["coherence"], runner="cloud"),
        agent("plan-risk", "Plan risk reviewer", _PLANNER, role=ROLE_PROMPTS["plan-risk"], runner="cloud"),
        agent("librarian", "Librarian", _WORKER, role=ROLE_PROMPTS["librarian"],
              grants=[_home_grant("librarian"), _WIKI_GRANT], env=env_open),
        agent("customer-support", "Customer support", _WORKER, role=ROLE_PROMPTS["customer-support"],
              mcp=["agentos", "inbox", "front"], env=env_none, inbox=True),
        agent("diagnostic", "Bug diagnostic", _WORKER, role=ROLE_PROMPTS["diagnostic"],
              mcp=["agentos", "inbox", "r2-fs", "github"], env=env_open),
        agent("linkedin-content", "LinkedIn content", _WORKER,
              role=ROLE_PROMPTS["linkedin-content"], env=env_none),
    ]

    templates = [
        {
            "project_id": project_id,
            "name": "compound-engineer-workflow",
            "description": "~3-hour fully managed feature build: spec → plan → review → revise → implement → code review → fixes → wiki → human PR review.",
            "variables": ["branchName", "featureTitle"],
            "steps": compound_engineer_steps(),
        }
    ]

    return {"agents": agents, "environments": environments, "skills": skills,
            "mcp_connections": mcp_connections, "templates": templates}


def compound_engineer_steps() -> list[dict]:
    return [
        {"name": "Write a spec", "agentName": "spec",
         "prompt": "Write a detailed feature specification for {{featureTitle}} (branch {{branchName}}). "
                   "Deliverable: a spec file attached to the task. Exit: spec attached, awaiting human approval — do not mark done.",
         "approvalGate": True},
        {"name": "Plan", "agentName": "plan",
         "prompt": "Turn the approved spec for {{featureTitle}} (branch {{branchName}}) into a concrete, ordered implementation plan. "
                   "Deliverable: a plan file attached to the task. Exit: plan attached, task finished. Do not implement.",
         "approvalGate": False},
        {"name": "Plan review", "agentName": "review-coordinator",
         "prompt": "Spawn the plan-review specialists for {{featureTitle}}. "
                   "Deliverable: consolidated report with must-fix / should-fix items. Exit: every specialist reported and consolidation attached.",
         "approvalGate": False},
        {"name": "Revise plan", "agentName": "plan",
         "prompt": "Revise the plan for {{featureTitle}} using the consolidated review. "
                   "Deliverable: revised plan attached. Exit: must-fix items addressed.",
         "approvalGate": False},
        {"name": "Implementation", "agentName": "implementation-plan-executioner",
         "prompt": "Implement {{featureTitle}} on branch {{branchName}} exactly per the revised plan. "
                   "Deliverable: committed implementation. Exit: all steps done, E2E run, work committed.",
         "approvalGate": False},
        {"name": "Code review", "agentName": "review-coordinator",
         "prompt": "Spawn code-review specialists for {{featureTitle}} on branch {{branchName}}. "
                   "Deliverable: consolidated report. Exit: every specialist reported.",
         "approvalGate": False},
        {"name": "Apply review fixes", "agentName": "senior-dev",
         "prompt": "Apply consolidated code-review fixes for {{featureTitle}} on branch {{branchName}}. "
                   "Deliverable: committed fixes. Exit: must-fix resolved, E2E passing.",
         "approvalGate": False},
        {"name": "Librarian", "agentName": "librarian",
         "prompt": "Update the internal wiki to reflect {{featureTitle}} on branch {{branchName}}. "
                   "Deliverable: updated wiki pages. Exit: wiki updated.",
         "approvalGate": False},
        {"name": "Human PR review", "agentName": "human",
         "prompt": "Review and merge the PR for {{featureTitle}} on branch {{branchName}}.",
         "approvalGate": True},
    ]


def bugfix_chain_steps() -> list[dict]:
    return [
        {"name": "Implement fix", "agentName": "implementation-plan-executioner",
         "prompt": "Implement the fix for {{featureTitle}} on branch {{branchName}}. Context: {{bugContext}}", "approvalGate": False},
        {"name": "Plan", "agentName": "plan",
         "prompt": "Turn the fix into a concrete plan for {{featureTitle}} on branch {{branchName}}.", "approvalGate": False},
        {"name": "Plan review", "agentName": "review-coordinator",
         "prompt": "Review the fix plan ({{featureTitle}}, branch {{branchName}}) and consolidate must-fix / should-fix.", "approvalGate": False},
        {"name": "Apply review fixes", "agentName": "senior-dev",
         "prompt": "Apply consolidated review fixes for {{featureTitle}} on branch {{branchName}}. Commit.", "approvalGate": False},
        {"name": "E2E verification", "agentName": "implementation-plan-executioner",
         "prompt": "Run E2E tests for {{featureTitle}} on branch {{branchName}}; fix failures; commit.", "approvalGate": False},
        {"name": "Human merge", "agentName": "human",
         "prompt": "Review and merge the bug fix for {{featureTitle}} on branch {{branchName}}.", "approvalGate": True},
    ]
